translate `Updating Arrays in State` (#460)
reactjs/tr.react.dev: src/content/learn/updating-arrays-in-state.md

---
title: State İçerisindeki Dizileri Güncelleme
---

<Intro>

JavaScript'te diziler mutasyona uğratılabilir (mutable) ancak siz state'te sakladığınız dizileri mutasyona uğratılamaz (immutable) olarak görmelisiniz. Tıpkı nesnelerde olduğu gibi, state'te saklanan bir diziyi güncellemek istediğiniz zaman yeni bir dizi oluşturmanız (veya var olanın bir kopyasını oluşturmanız) ve ardından yeni oluşturduğunuz diziyi kullanmak için state'i güncellemeniz gerekmektedir.

</Intro>

<YouWillLearn>

- React state'indeki bir diziye öğeler nasıl eklenir, çıkarılır ya da değiştirilir
- Bir dizi içindeki nesne nasıl güncellenir
- Immer kullanarak dizi kopyalama işlemi daha az tekrarla nasıl yapılır

</YouWillLearn>

## Dizileri mutasyona uğratmadan güncelleme {/*updating-arrays-without-mutation*/}

> JavaScript'te diziler de bir tür nesnedir. [Nesnelerde olduğu gibi](/learn/updating-objects-in-state), **React state'indeki dizileri salt okunur olarak görmelisiniz.** Bu, `arr[0] = 'bird'` şeklinde bir dizi içindeki öğeleri başka değerlere yeniden atamamanız, ayrıca `push()` ve `pop()` gibi dizileri mutasyona uğratan JavaScript metodlarını kullanmamanız gerektiği anlamına gelir.

Reviewer: ```suggestion
JavaScript'te diziler bir nesne türüdür. [Nesnelerde olduğu gibi](/learn/updating-objects-in-state), **React state'indeki dizileri salt okunur olarak görmelisiniz.** Bu, `arr[0] = 'bird'` şeklinde bir dizi içindeki öğeleri başka değerlere yeniden atamamanız, `push()` ve `pop()` gibi dizileri mutasyona uğratan JavaScript metodlarını kullanmamanız gerektiği anlamına gelir.
```

Aşağıda sık kullanılan dizi metodları tablo halinde gösterilmiştir. React state'indeki dizilerle çalışırken sol sütundaki metodları kullanmaktan kaçınarak sağ sütundaki metodları tercih etmelisiniz.

|           | kaçınılacaklar (diziyi mutasyona uğratır)           |  tercih edilecekler (yeni bir dizi döndürür)                                        |
| --------- | ----------------------------------- | ------------------------------------------------------------------- |
| eklemek    | `push`, `unshift`                   | `concat`, `[...arr]` spread sözdizimi ([örnek](#adding-to-an-array)) |
| çıkartmak  | `pop`, `shift`, `splice`            | `filter`, `slice` ([örnek](#removing-from-an-array))              |
| değiştirmek | `splice`, `arr[i] = ...` ataması | `map` ([örnek](#replacing-items-in-an-array))                     |
| sıralamak   | `reverse`, `sort`                   | ilk önce diziyi kopyalayın ([örnek](#making-other-changes-to-an-array)) |

> Alternatif olarak, her iki sütundaki metodları kullanmanıza izin veren [Immer'ı] (#write-concise-update-logic-with-immer) tercih edebilirsiniz.

Reviewer: ```suggestion
Alternatif olarak, her iki sütundaki metodları kullanmanıza izin veren [Immer'ı](#write-concise-update-logic-with-immer) tercih edebilirsiniz.
```

<Pitfall>

Ne yazık ki, [`slice`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice) ve [`splice`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice) metodları isim olarak benzeseler bile birbirlerinden çok farklıdırlar.

* `slice` metodu dizinin bir parçasını veya tamamını kopyalamınızı sağlar.
* `splice` diziyi **mutasyona uğratır** (diziye yeni öğe eklemek ya da var olanı çıkartmak için kullanılır).

React'te, `slice` (`p` yok!) metodunu daha sık kullanacaksınız çünkü state'teki nesneleri veya dizileri mutasyona uğratmak istemezsiniz. [Nesneleri Güncelleme](/learn/updating-objects-in-state) sayfasında mutasyon nedir ve state için neden kullanılmamalıdır öğrenebilirsiniz.

</Pitfall>

> ### Bir diziye eklemek {/*adding-to-an-array*/}

Reviewer: ```suggestion
### Diziye öğe ekleme {/*adding-to-an-array*/}
```

> `push()` metodu diziyi mutasyona uğratacaktır ancak biz bunu istemiyoruz.

Reviewer: React dökümanının genel kuralı sizli konuşmak. Biz dilini kullanmamamız lazım:

```suggestion
`push()` metodu diziyi mutasyona uğratacaktır, ki bunu istemezsiniz:
```

</Sandpack>

> Bunun yerine, mevcut öğeleri *ve* dizinin son elemanı olarak yeni öğeyi içeren *yeni* diziyi oluşturun. Bunu yapmanın birden çok yolu vardır ancak en kolay yol `...` [dizi spread](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#spread_in_array_literals) sözdizimini kullanmaktır:

Reviewer: ```suggestion
Bunun yerine, mevcut öğeleri *ve* son eleman olarak yeni öğeyi içeren *yeni* diziyi oluşturun. Bunu yapmanın birden çok yolu vardır ancak en kolay yol `...` [dizi spread](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#spread_in_array_literals) sözdizimini kullanmaktır:
```

```js

> setArtists( // State'i yeni bir dizi 
>   [ // ile değiştirin
>     ...artists, // bu eski öğelerin hepsini içerir
>     { id: nextId++, name: name } // ve dizinin sonuna eklenecek yeni öğe

Reviewer: ```suggestion
setArtists( // State'i yeni bir dizi 
  [ // ile değiştirin.
    ...artists, // Bu eski öğelerin tümünü
    { id: nextId++, name: name } // ve sona eklenecek yeni öğeyi içerir.
```

</Sandpack>

Dizi spread sözdizimi yeni öğeyi orijinal `...artists`'den *önceye* yerleştirerek dizinin başına eklemenizi de sağlar:

```js
setArtists([
  { id: nextId++, name: name },
  ...artists // Eski öğeleri dizinin sonuna yerleştir
]);
```

Bu şekilde, spread sözdizimi `push()` (öğeyi dizinin sonuna eklemek) ve `unshift()` (öğeyi dizinin başına eklemek) metodlarının görevini yapabilir. Yukarıdaki sandbox'ta deneyebilirsiniz!

> ### Bir diziden çıkartmak {/*removing-from-an-array*/}

Reviewer: ```suggestion
### Diziden öğe çıkartma {/*removing-from-an-array*/}
```

> Diziden bir öğeyi çıkartmanın en kolay yolu o öğeyi *filtrelemek'tir.* Bir başka deyişle, o öğeyi içermeyen yeni bir dizi oluşturmaktır. Bunu yapmak için `filter` metodunu kullanabilirsiniz. Örneğin:

Reviewer: ```suggestion
Diziden bir öğeyi çıkartmanın en kolay yolu o öğeyi *filtrelemektir.* Bir başka deyişle, o öğeyi içermeyen yeni bir dizi oluşturmaktır. Bunu yapmak için `filter` metodunu kullanabilirsiniz. Örneğin:
```

</Sandpack>

"Sil" butonuna birkaç kez tıklayın ve tıklama yöneticisine bakın.

```js
setArtists(
  artists.filter(a => a.id !== artist.id)
);
```

> Burada `artists.filter(a => a.id !== artist.id)` ifadesi "ID'leri `artist.id`'den farklı olan `artists`'den oluşan yeni bir dizi oluştur" anlamına gelmektedir. Diğer bir deyişle, her bir artist'e karşılık gelen "Sil" butonu o artist'i diziden filtreleyecek ve nihai dizi ile yeniden render isteği gönderecektir. `filter` metodunun orijinal diziyi değiştirmediğini unutmayın.

Reviewer: ```suggestion
Burada `artists.filter(a => a.id !== artist.id)` ifadesi "`artists` dizisini kullanarak ID'leri `artist.id`'den farklı olan öğelerle yeni bir dizi oluştur" anlamına gelmektedir. Diğer bir deyişle, her bir artist'e karşılık gelen "Sil" butonu o artist'i diziden filtreleyecek ve nihai dizi ile yeniden render isteği gönderecektir. `filter` metodunun orijinal diziyi değiştirmediğini unutmayın.
```

> ### Bir diziyi dönüştürmek {/*transforming-an-array*/}

Reviewer: ```suggestion
### Diziyi dönüştürme {/*transforming-an-array*/}
```

> Eğer dizideki bazı ya da tüm öğeleri değiştirmek isterseniz **yeni** bir dizi oluşturmak için `map()` metodunu kullanabilirsiniz. `map`'e ileteceğiniz fonksiyon, verisine veya indeksine (veya her ikisine) dayalı olarak her bir öğeyle ne yapacağınızı belirler.

Reviewer: ```suggestion
Dizideki bazı ya da tüm öğeleri değiştirmek isterseniz **yeni** bir dizi oluşturmak için `map()` metodunu kullanabilirsiniz. `map`'e ileteceğiniz fonksiyon, verisine veya indeksine (veya her ikisine) bağlı olarak her bir öğeyle ne yapacağınızı belirler.
```

</Sandpack>

> ### Bir dizideki öğeleri değiştirmek {/*replacing-items-in-an-array*/}

Reviewer: ```suggestion
### Dizideki öğeleri değiştirme {/*replacing-items-in-an-array*/}
```

</Sandpack>

> ### Dizide belli bir konuma eklemek {/*inserting-into-an-array*/}

Reviewer: ```suggestion
### Dizideki belirli konuma öğe ekleme {/*inserting-into-an-array*/}
```

> Bazen bir öğeyi ne dizinin başına ne de dizinin sonuna değil de belirlediğiniz başka bir konuma eklemek isteyebilirsiniz. Bunu yapmak için, `...` dizi spread sözdizimini `slice()` metodu ile beraber kullanabilirsiniz. `slice()` metodu diziden bir "dilim (slice)" almanızı sağlar. Yeni öğeyi eklemek için, eklemek istediğiniz konumdan önceki dilimden spread sözdizimi ile yeni bir dizi oluşturacak, ardından yeni öğeyi ekleyecek ve son olarak da orijinal dizinin geri kalanını ekleyeceksiniz.

Reviewer: ```suggestion
Bazen bir öğeyi dizinin başına ya da sonuna değil de belirli bir bir konuma eklemek isteyebilirsiniz. Bunu yapmak için, `...` dizi spread sözdizimini `slice()` metodu ile beraber kullanabilirsiniz. `slice()` metodu diziden bir "dilim (slice)" almanızı sağlar. Yeni öğeyi eklemek için, eklemek istediğiniz konumdan önceki dilimden spread sözdizimi ile yeni bir dizi oluşturacak, ardından yeni öğeyi ekleyecek ve son olarak da orijinal dizinin geri kalanını ekleyeceksiniz.
```

> Bu örnekte, Ekle butonu her zaman `1` indeksinde ekler:

Reviewer: ```suggestion
Bu örnekte, Ekle butonu her zaman `1.` indekse ekler:
```

</Sandpack>

> ### Bir dizide başka değişiklikler yapmak {/*making-other-changes-to-an-array*/}

Reviewer: ```suggestion
### Dizide başka değişiklikler yapma {/*making-other-changes-to-an-array*/}
```

> Spread sözdizimi, `map()` ve `filter()` metodları gibi diziyi mutasyona uğratmayan yöntemlerle yapamayacağınız bazı şeyler vardır. Örneğin, bir diziyi sıralamak veya dizi öğelerini ters çevirmek isteyebilirsiniz. JavaScript'in `reverse()` ve `sort()` metodları orijinal diziyi mutasyona uğrattığından dolayı direkt olarak bu metodları kullanamazsınız.

Reviewer: ```suggestion
Spread sözdizimi veya `map()` ve `filter()` gibi diziyi mutasyona uğratmayan metodlarla yapamayacağınız bazı şeyler vardır. Örneğin, bir diziyi sıralamak veya dizi öğelerini ters çevirmek isteyebilirsiniz. JavaScript'in `reverse()` ve `sort()` metodları orijinal diziyi değiştirir, bu nedenle doğrudan kullanamazsınız.
```

Ancak, **bir diziyi kopyalasanız bile dizinin  _içindeki_ öğeleri doğrudan mutasyona uğratamazsınız.** Bunun nedeni yaptığınız kopyalamanın yüzeysel (shallow) olmasıdır. Yani yeni dizi, orijinal diziyle aynı öğeleri içermektedir. Dolayısıyla, kopyalanan dizinin içindeki bir nesneyi değiştirdiğiniz zaman mevcut state'i de mutasyona uğratmış olursunuz. Örneğin, aşağıdaki gibi bir kod sorunludur. 

```js
const nextList = [...list];
nextList[0].seen = true; // Sorun: list[0]'ı mutasyona uğratır
setList(nextList);
```

`nextList` ve `list` iki farklı dizi olmasına rağmen, **`nextList[0]` ve `list[0]` ifadeleri aynı nesneyi işaret eder.** Yani `nextList[0].seen` değerini değiştirirseniz, aynı zamanda `list[0].seen` değerini de değiştirmiş olursunuz. Bu, state'i mutasyona uğratmaktır ki bundan kaçınmalısınız! Bu sorunu [iç içe JavaScript nesnelerini güncelleme](/learn/updating-objects-in-state#updating-a-nested-object) yöntemine benzer şekilde, değiştirmek istediğiniz öğeleri mutasyona uğratmak yerine tek tek kopyalayarak çözebilirsiniz. Nasıl yapıldığını görelim.

> ## Dizi içindeki nesneleri güncelleme {/*updating-objects-inside-arrays*/}

Reviewer: ```suggestion
## Dizideki nesneleri güncelleme {/*updating-objects-inside-arrays*/}
```

> Nesneler dizilerin  _gerçekten_ "içinde" değillerdir. Yazdığınız kodda "içeride" olarak görünebilir ancak bir dizideki her nesne, dizinin "işaret ettiği" ayrı bir değerdir. Bu yüzden `list[0]` gibi iç içe ifadeleri değiştirirken dikkatli olmalısınız. Başka bir kişinin sanat eseri listesi (artwork list), dizinin aynı öğesine işaret edebilir!

Reviewer: ```suggestion
Nesneler  _gerçekte_ dizilerin "içinde" yer almazlar. Yazdığınız kodda "içinde" gibi görünebilir ancak bir dizideki her nesne, dizinin "işaret ettiği" ayrı bir değerdir. Bu yüzden `list[0]` gibi iç içe ifadeleri değiştirirken dikkatli olmalısınız. Başka bir kişinin sanat eseri listesi (artwork list), dizinin aynı öğesine işaret edebilir!
```

**İç içe geçmiş state'i güncellerken, güncellemek istediğiniz noktadan en üst düzeye kadar kopyalar oluşturmanız gerekir.** Şimdi bunun nasıl olduğunu görelim.

> Bu örnekte, iki farklı sanat eseri listesi aynı ilk state'e (initialList) sahiptir. Bu listelerin izole olmaları gerekirdi ancak bir mutasyon nedeniyle yanlışlıkla state'leri paylaşmaktadırlar ve listedeki bir kutuyu işaretlemek diğer listedeki kutuları da etkilemektedir:

Reviewer: ```suggestion
Bu örnekte, iki farklı sanat eseri listesi aynı başlangıç state'ine sahiptir. Bu listelerin izole olmaları gerekirdi ancak bir mutasyon nedeniyle yanlışlıkla state'leri paylaşmaktadırlar ve listedeki bir kutuyu işaretlemek diğer listeyi de etkilemektedir:
```

});
```

Bunun nedeni, _orijinal_ state'i mutasyona uğratmamanızdır. Burada Immer tarafından sağlanan özel bir `draft` nesnesini mutasyona uğratmaktayız. Benzer şekilde, `draft` nesnesine `push()` ve `pop()` gibi mutasyona neden olan metodları da uygulayabilirsiniz.

Arka planda Immer, `draft`'a yaptığınız değişikliklere göre her zaman bir sonraki state'i sıfırdan oluşturur. Bu, olay yönetecilerinizi, state'i hiç mutasyona uğratmadan kısa ve öz olarak tutar.

<Recap>

- State'e dizi koyabilirsiniz ancak değiştiremezsiniz.
- Bir diziyi mutasyona uğratmak yerine o dizinin *yeni* bir sürümünü oluşturun ve state'i buna göre güncelleyin.
- `[...arr, newItem]` dizi spread sözdizimini kullanarak yeni öğelerle bir dizi oluşturabilirsiniz
- `filter()` ve `map()` metodlarını kullanarak filtrelenmiş ya da dönüştürülmüş yeni diziler oluşturabilirsiniz.
- Kodunuzu kısa ve öz tutmak için Immer'ı kullanabilirsiniz.

</Recap>

<Challenges>

#### Alışveriş sepetindeki ürünü güncelleyin {/*update-an-item-in-the-shopping-cart*/}

> `handleIncreaseClick` fonksiyonunun mantığını yazın ki "+" butonuna tıklandığında ürünün sepetteki sayısı artsın:

Reviewer: ```suggestion
"+" butonuna tıklandığında ilgili sayının artması için `handleIncreaseClick` mantığını doldurun:
```

</Solution>

#### Immer kullanarak mutasyonları düzeltin {/*fix-the-mutations-using-immer*/}

> Bu örnekle bir önceki örnek aynıdır. Bu sefer mutasyonları Immer kullanarak düzelteceğiz. `useImmer` halihazırda import edilmiştir, yani kullanmak için `todos` state değişkenini değiştirmeniz gerekmektedir.

Reviewer: ```suggestion
Bu örnekle bir önceki örnek aynıdır.  Bu sefer Immer kullanarak mutasyonları düzeltin. Kolaylık sağlaması için, `useImmer` halihazırda içe aktarılmıştır. Bunu kullanarak`todos` state değişkenini değiştirmeniz gerekir.
```
